File: apps/web/src/lib/research.ts
export type MarketStats = {
  source: string;
  symbol: string;
  timeframe: string;
  last_price: number | null;
  change_1h: number | null;
  change_24h: number | null;
  change_7d: number | null;
  change_30d: number | null;
  high_24h: number | null;
  low_24h: number | null;
  volume_24h: number | null;
  realized_vol_24h: number | null;
  realized_vol_7d: number | null;
  bar_count: number;
  earliest_ts: string | null;
  latest_ts: string | null;
};

export type CorrelationOut = {
  source: string;
  timeframe: string;
  days: number;
  symbols: string[];
  matrix: (number | null)[][];
  samples: number[][];
  computed_at: string;
};

export function formatPct(x: number | null | undefined, digits = 2): string {
  if (x == null || !Number.isFinite(x)) return "—";
  const sign = x > 0 ? "+" : "";
  return `${sign}${(x * 100).toFixed(digits)}%`;
}

export function formatPrice(x: number | null | undefined): string {
  if (x == null || !Number.isFinite(x)) return "—";
  if (x >= 1000) return x.toLocaleString(undefined, { maximumFractionDigits: 2 });
  if (x >= 1) return x.toFixed(4);
  return x.toPrecision(4);
}

export type FundingPoint = {
  ts: string;
  rate: number;
};

export type FundingHistoryOut = {
  source: string;
  symbol: string;
  days: number;
  points: FundingPoint[];
  mean: number | null;
  annualized: number | null;
};

export function formatCompactNumber(x: number | null | undefined): string {
  if (x == null || !Number.isFinite(x)) return "—";
  return new Intl.NumberFormat(undefined, {
    notation: "compact",
    maximumFractionDigits: 2,
  }).format(x);
}

File: apps/api/src/maelstrom_api/schemas/research.py
from datetime import datetime

from pydantic import BaseModel, Field


class MarketStats(BaseModel):
    source: str
    symbol: str
    timeframe: str
    last_price: float | None = None
    change_1h: float | None = None
    change_24h: float | None = None
    change_7d: float | None = None
    change_30d: float | None = None
    high_24h: float | None = None
    low_24h: float | None = None
    volume_24h: float | None = None
    realized_vol_24h: float | None = None  # annualized, as a fraction (0.65 = 65%)
    realized_vol_7d: float | None = None
    bar_count: int = 0
    earliest_ts: datetime | None = None
    latest_ts: datetime | None = None


class CorrelationRequest(BaseModel):
    source: str
    symbols: list[str] = Field(min_length=2, max_length=20)
    timeframe: str = "1h"
    days: int = Field(default=30, ge=1, le=365)


class CorrelationOut(BaseModel):
    source: str
    timeframe: str
    days: int
    symbols: list[str]
    matrix: list[list[float | None]]
    # Number of overlapping return samples per pair (square, symmetric).
    samples: list[list[int]]
    computed_at: datetime
